exporter/containerimage/writer.go

```diff
@@ -333,6 +409,126 @@ func (ic *ImageWriter) commitDistributionManifest(ctx context.Context, ref cache
+	}
+
+	layers := []ocispecs.Descriptor{}
+	for _, statement := range statements {
+		data, err := json.Marshal(statement)
+		if err != nil {
+			return nil, err
+		}
+
+		digest := digest.FromBytes(data)
+		desc := ocispecs.Descriptor{
+			MediaType: attestationType,
+			Digest:    digest,
+			Size:      int64(len(data)),
+			Annotations: map[string]string{
+				"containerd.io/uncompressed": digest.String(),
+				"in-toto.io/predicate-type":  statement.PredicateType,
+			},
+		}
+		done := oneOffProgress(ctx, "exporting attestation "+digest.String())
+		if err := content.WriteBlob(ctx, ic.opt.ContentStore, digest.String(), bytes.NewReader(data), desc); err != nil {
+			return nil, done(errors.Wrapf(err, "error writing data blob %s", digest))
+		}
+		done(nil)
+
+		layers = append(layers, desc)
+	}
+
+	config, err := attestationsConfig(layers)
+	if err != nil {
+		return nil, err
+	}
+	configDigest := digest.FromBytes(config)
+
+	mfst := struct {
+		// MediaType is reserved in the OCI spec but
+		// excluded from go types.
+		MediaType string `json:"mediaType,omitempty"`
+
+		ocispecs.Manifest
+	}{
+		MediaType: manifestType,
+		Manifest: ocispecs.Manifest{
+			Versioned: specs.Versioned{
+				SchemaVersion: 2,
+			},
+			Config: ocispecs.Descriptor{
+				Digest:    configDigest,
+				Size:      int64(len(config)),
+				MediaType: configType,
+			},
+		},
+	}
+
+	labels := map[string]string{
+		"containerd.io/gc.ref.content.0": configDigest.String(),
+	}
+	for i, desc := range layers {
+		removeLayerAnnotations(&desc, oci)
+		mfst.Layers = append(mfst.Layers, desc)
+		labels[fmt.Sprintf("containerd.io/gc.ref.content.%d", i+1)] = desc.Digest.String()
+	}
+
+	mfstJSON, err := json.MarshalIndent(mfst, "", "   ")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal manifest")
+	}
+
+	mfstDigest := digest.FromBytes(mfstJSON)
+	mfstDesc := ocispecs.Descriptor{
+		Digest: mfstDigest,
+		Size:   int64(len(mfstJSON)),
+	}
+	mfstDone := oneOffProgress(ctx, "exporting attestation manifest "+mfstDigest.String())
+
+	if err := content.WriteBlob(ctx, ic.opt.ContentStore, mfstDigest.String(), bytes.NewReader(mfstJSON), mfstDesc, content.WithLabels((labels))); err != nil {
+		return nil, mfstDone(errors.Wrapf(err, "error writing manifest blob %s", mfstDigest))
+	}
+	mfstDone(nil)
+
+	configDesc := ocispecs.Descriptor{
+		Digest:    configDigest,
+		Size:      int64(len(config)),
+		MediaType: configType,
+	}
+	configDone := oneOffProgress(ctx, "exporting config "+configDigest.String())
+
+	if err := content.WriteBlob(ctx, ic.opt.ContentStore, configDigest.String(), bytes.NewReader(config), configDesc); err != nil {
+		return nil, configDone(errors.Wrap(err, "error writing config blob"))
+	}
+	configDone(nil)
+
+	return &ocispecs.Descriptor{
+		Digest:    mfstDigest,
+		Size:      int64(len(mfstJSON)),
+		MediaType: manifestType,
+		Annotations: map[string]string{
+			annotationReferenceType:        "buildkit-attestations",
```

Comment:
I think this should be "attestation-manifest". Buildkit is just one of the generators.